File: go/transport.go
package epoch

// Transport defines the minimal send/poll/close contract.
type Transport interface {
	Send(message Message)
	Poll(max int) []Message
	Close()
}

type InMemoryTransport struct {
	queue []Message
}

func (t *InMemoryTransport) Send(message Message) {
	t.queue = append(t.queue, message)
}

func (t *InMemoryTransport) Poll(max int) []Message {
	if max <= 0 || len(t.queue) == 0 {
		return nil
	}
	if max > len(t.queue) {
		max = len(t.queue)
	}
	out := append([]Message(nil), t.queue[:max]...)
	t.queue = t.queue[max:]
	return out
}

func (t *InMemoryTransport) Close() {
	t.queue = nil
}

File: dotnet/Transport.cs
namespace Epoch;

public interface ITransport
{
    void Send(Message message);
    List<Message> Poll(int max);
    void Close();
}

public sealed class InMemoryTransport : ITransport
{
    private readonly Queue<Message> _queue = new();

    public void Send(Message message)
    {
        _queue.Enqueue(message);
    }

    public List<Message> Poll(int max)
    {
        var results = new List<Message>();
        while (_queue.Count > 0 && results.Count < max)
        {
            results.Add(_queue.Dequeue());
        }
        return results;
    }

    public void Close()
    {
        _queue.Clear();
    }
}
